[Bug][Jobs] Job submission CLI didn't work on first ray install

### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.


### Ray Component

Ray Serve

### What happened + What you expected to happen

Assuming user starts from clean conda env and install ray packages via pip, job submission CLI doesn't work.

### Versions / Dependencies

on master and nightly 

### Reproduction script

Basically create new conda, install nightly, install ray default, ensure we have dashboard module running, then submit simple things with CLI

```
conda create --name cli-jobs python=3.8

conda activate cli-jobs

pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp38-cp38-macosx_10_15_x86_64.whl

python -c "import ray;print(print(ray.__commit__))"

pip install `ray[default]`

ray start --head
export RAY_ADDRESS="http://127.0.0.1:8265"
ray job submit -- "echo hello"
```

We will repeatedly get 
```

❯ ray job submit -- "echo hello"
Traceback (most recent call last):
  File "/Users/jiaodong/anaconda3/envs/repro/bin/ray", line 8, in <module>
    sys.exit(main())
  File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/ray/scripts/scripts.py", line 1989, in main
    return cli()
  File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/ray/dashboard/modules/job/cli.py", line 99, in job_submit
    job_id = client.submit_job(
  File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/ray/dashboard/modules/job/sdk.py", line 213, in submit_job
    return JobSubmitResponse(**r.json()).job_id
TypeError: __init__() got an unexpected keyword argument 'result'
```

Where from dashboard module logs it suggests:

```
 64 2021-11-23 12:49:40,808 ERROR utils.py:132 -- Handle POST /api/jobs/ failed.
 65 Traceback (most recent call last):
 66   File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/ray/dashboard/utils.py", line 130, in _handler_route
 67     return await handler(bind_info.instance, req)
 68   File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/ray/dashboard/modules/job/job_head.py", line 39, in check
 69     ray.init(address="auto", namespace=RAY_INTERNAL_JOBS_NAMESPACE)
 70   File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
 71     return func(*args, **kwargs)
 72   File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/ray/worker.py", line 755, in init
 73     builder = ray.client(address, _deprecation_warn_enabled=False)
 74   File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/ray/client_builder.py", line 347, in client
 75     builder = _get_builder_from_address(address)
 76   File "/Users/jiaodong/anaconda3/envs/repro/lib/python3.8/site-packages/ray/client_builder.py", line 318, in _get_builder_from_address
 77     assert "ClientBuilder" in dir(module), (f"Module: {module_string} does "
 78 AssertionError: Module: http does not have ClientBuilder.
```

Note that it will work again if we try to re-start dashboard from another terminal.

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

--- a/dashboard/modules/job/cli.py
+++ b/dashboard/modules/job/cli.py
@@ -14,6 +14,7 @@ logger = logging.getLogger(__name__)
 def _get_sdk_client(address: Optional[str],
                     create_cluster_if_needed: bool = False
                     ) -> JobSubmissionClient:
+
     if address is None:
         if "RAY_ADDRESS" not in os.environ:
             raise ValueError(
@@ -21,6 +22,7 @@ def _get_sdk_client(address: Optional[str],
                 "or RAY_ADDRESS environment variable.")
         address = os.environ["RAY_ADDRESS"]
 
+    logger.info(f"Creating JobSubmissionClient at address: {address}")
     return JobSubmissionClient(address, create_cluster_if_needed)
 
 

--- a/dashboard/modules/job/job_head.py
+++ b/dashboard/modules/job/job_head.py
@@ -27,6 +27,8 @@ from ray.dashboard.modules.job.common import (
 from ray.dashboard.modules.job.job_manager import JobManager
 
 logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
 routes = dashboard_utils.ClassMethodRouteTable
 
 RAY_INTERNAL_JOBS_NAMESPACE = "_ray_internal_jobs"
@@ -35,9 +37,21 @@ RAY_INTERNAL_JOBS_NAMESPACE = "_ray_internal_jobs"
 def _init_ray_and_catch_exceptions(f: Callable) -> Callable:
     @wraps(f)
     async def check(self, *args, **kwargs):
-        if not ray.is_initialized():
-            ray.init(address="auto", namespace=RAY_INTERNAL_JOBS_NAMESPACE)
         try:
+            if not ray.is_initialized():
+                try:
+                    address = self._redis_address
+                    redis_pw = self._redis_password
+                    logger.info(f"Connecting to ray with address={address}, "
+                                f"redis_pw={redis_pw}")
+                    ray.init(
+                        address=address,
+                        namespace=RAY_INTERNAL_JOBS_NAMESPACE,
+                        _redis_password=redis_pw)
+                except Exception as e:
+                    ray.shutdown()
+                    raise e from None
+
             return await f(self, *args, **kwargs)
         except Exception as e:
             logger.exception(f"Unexpected error in handler: {e}")
@@ -52,6 +66,9 @@ class JobHead(dashboard_utils.DashboardHeadModule):
     def __init__(self, dashboard_head):
         super().__init__(dashboard_head)
 
+        ip, port = dashboard_head.redis_address
+        self._redis_address = f"{ip}:{port}"
+        self._redis_password = dashboard_head.redis_password
         self._job_manager = None
 
     async def _parse_and_validate_request(self, req: Request,

--- a/dashboard/modules/job/tests/test_cli.py
+++ b/dashboard/modules/job/tests/test_cli.py
@@ -6,6 +6,7 @@ import tempfile
 import os
 from unittest import mock
 import yaml
+from typing import Optional
 
 from click.testing import CliRunner
 
@@ -47,11 +48,17 @@ def runtime_env_formats():
 
 
 @contextmanager
-def set_env_var(key: str, val: str):
+def set_env_var(key: str, val: Optional[str] = None):
     old_val = os.environ.get(key, None)
-    os.environ[key] = val
+    if val is not None:
+        os.environ[key] = val
+    elif key in os.environ:
+        del os.environ[key]
+
     yield
-    del os.environ[key]
+
+    if key in os.environ:
+        del os.environ[key]
     if old_val is not None:
         os.environ[key] = old_val
 

--- a/dashboard/modules/job/tests/test_cli_integration.py
+++ b/dashboard/modules/job/tests/test_cli_integration.py
@@ -1,0 +1,91 @@
+from contextlib import contextmanager
+import logging
+import sys
+import subprocess
+
+import pytest
+
+from ray.dashboard.modules.job.tests.test_cli import set_env_var
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture
+def ray_start_stop():
+    subprocess.check_output(["ray", "start", "--head"])
+    yield
+    subprocess.check_output(["ray", "stop", "--force"])
+
+
+@contextmanager
+def ray_cluster_manager():
+    """
+    Used not as fixture in case we want to set RAY_ADDRESS first.
+    """
+    subprocess.check_output(["ray", "start", "--head"])
+    yield
+    subprocess.check_output(["ray", "stop", "--force"])
+
+
+class TestSubmitIntegration:
+    """
+    Integration version of job CLI test that ensures interaction with the
+    following components are working as expected:
+
+    1) Ray client: use of RAY_ADDRESS and ray.init() in job_head.py
+    2) Ray dashboard: `ray start --head`
+    """
+
+    def test_empty_ray_address(self, ray_start_stop):
+        with set_env_var("RAY_ADDRESS", None):
+            completed_process = subprocess.run(
+                ["ray", "job", "submit", "--", "echo hello"],
+                stderr=subprocess.PIPE)
+            stderr = completed_process.stderr.decode("utf-8")
+            # Current dashboard module that raises no exception from requests..
+            assert ("Address must be specified using either the "
+                    "--address flag or RAY_ADDRESS environment") in stderr
+
+    def test_ray_client_adress(self, ray_start_stop):
+        with set_env_var("RAY_ADDRESS", "127.0.0.1:8265"):
+            completed_process = subprocess.run(
+                ["ray", "job", "submit", "--", "echo hello"],
+                stderr=subprocess.PIPE)
+            stderr = completed_process.stderr.decode("utf-8")
+            # Current dashboard module that raises no exception from requests..
+            assert "Query the status of the job" in stderr
+
+    def test_valid_http_ray_address(self, ray_start_stop):
+        with set_env_var("RAY_ADDRESS", "http://127.0.0.1:8265"):
+            completed_process = subprocess.run(
+                ["ray", "job", "submit", "--", "echo hello"],
+                stderr=subprocess.PIPE)
+            stderr = completed_process.stderr.decode("utf-8")
+            # Current dashboard module that raises no exception from requests..
+            assert "Query the status of the job" in stderr
+
+    def test_set_ray_http_address_first(self):
+        with set_env_var("RAY_ADDRESS", "http://127.0.0.1:8265"):
+            with ray_cluster_manager():
+                completed_process = subprocess.run(
+                    ["ray", "job", "submit", "--", "echo hello"],
+                    stderr=subprocess.PIPE)
+                stderr = completed_process.stderr.decode("utf-8")
+                # Current dashboard module that raises no exception from
+                # requests..
+                assert "Query the status of the job" in stderr
+
+    def test_set_ray_client_address_first(self):
+        with set_env_var("RAY_ADDRESS", "127.0.0.1:8265"):
+            with ray_cluster_manager():
+                completed_process = subprocess.run(
+                    ["ray", "job", "submit", "--", "echo hello"],
+                    stderr=subprocess.PIPE)
+                stderr = completed_process.stderr.decode("utf-8")
+                # Current dashboard module that raises no exception from
+                # requests..
+                assert "Query the status of the job" in stderr
+
+
+if __name__ == "__main__":
+    sys.exit(pytest.main(["-v", __file__]))

--- a/dashboard/modules/job/tests/test_cli_integration.py
+++ b/dashboard/modules/job/tests/test_cli_integration.py
@@ -9,6 +9,7 @@ import pytest
 
 logger = logging.getLogger(__name__)
 
+
 @contextmanager
 def set_env_var(key: str, val: Optional[str] = None):
     old_val = os.environ.get(key, None)

--- a/dashboard/modules/job/tests/test_cli_integration.py
+++ b/dashboard/modules/job/tests/test_cli_integration.py
@@ -1,13 +1,28 @@
 from contextlib import contextmanager
+import os
 import logging
 import sys
 import subprocess
+from typing import Optional
 
 import pytest
 
-from ray.dashboard.modules.job.tests.test_cli import set_env_var
-
 logger = logging.getLogger(__name__)
+
+@contextmanager
+def set_env_var(key: str, val: Optional[str] = None):
+    old_val = os.environ.get(key, None)
+    if val is not None:
+        os.environ[key] = val
+    elif key in os.environ:
+        del os.environ[key]
+
+    yield
+
+    if key in os.environ:
+        del os.environ[key]
+    if old_val is not None:
+        os.environ[key] = old_val
 
 
 @pytest.fixture

